model-optimizer/mo/front/common/partial_infer/concat.py

```diff
@@ -37,33 +35,34 @@ def concat_infer(node):
     mask[axis] = True  # pylint: disable=unsupported-assignment-operation
     not_mask = np.logical_not(mask)  # pylint: disable=assignment-from-no-return
     for s in shapes[1:]:
-        s = int64_array(s)
-        if np.all(shape[not_mask] == s[not_mask]):  # TODO handle -1 in a special way
+        s = shape_array(s)
+        if np.ma.allequal(shape[not_mask], s[not_mask]):
             shape[mask] += s[mask]
         else:
-            log.error('Concat input shapes do not match')
-            return
+            raise Error('Concat input shapes do not match for node "{}" with axis {}'.format(node_name, axis))
 
-    node.out_node(0).shape = shape
-    if len(shape) != 4:
-        # exclude it from NHWC to NCHW conversion
-        if 'axis' in node.dim_attrs:
-            node.dim_attrs.remove('axis')
+    #  dynamic dimensions in the output (except the concat axis) can be deduced from input shape
+    for pos in range(len(shape)):
+        if shape[pos] is dynamic_dimension and pos != axis:
+            for in_shape in shapes:
+                if in_shape[pos] is not dynamic_dimension:
+                    shape[pos] = in_shape[pos]
 
+    node.out_port(0).data.set_shape(shape)
     PermuteAttrs.create_permute_attrs(node, attrs=[('axis', 'input:0')])
 
     values = [node.in_node(i).value for i in range(N)]
-    if any(v is None for v in values):
+    if any([v is None for v in values]):
         return
 
-    node.out_node(0).value = np.concatenate(values, axis=node.axis).astype(values[0].dtype, copy=False)
-    node.out_node(0).shape = np.array(node.out_node(0).value.shape, dtype=np.int64)
+    # if one of the input values are dynamic then we need to properly keep data type of the input
```

Comment:
```suggestion
    # if one of the input values are dynamic, you need to properly keep data type of the input
```

Reply:
This is a comment for the MO developer (not for the customer). I have re-phrased it.